WordPress/gutenberg
# Manage CSS for Global Styles

Continues work from: https://github.com/WordPress/gutenberg/pull/19883 https://github.com/WordPress/gutenberg/pull/20047
See larger picture at https://github.com/WordPress/gutenberg/issues/22296

This PR explores a way in which we can "manage CSS" for global styles. Instead of writing CSS directly, themes would provide their design choices for blocks. Gutenberg will generate a CSS that merges core, themes, and user preferences into one.

## How this works?

Let's take an existing pattern such as "Hero Two Columns" and try to convert it to "global styles format":

![Screenshot from 2020-05-14 10-53-26](https://user-images.githubusercontent.com/583546/81914125-2afabb80-95d1-11ea-9852-37bb970c4fd1.png)

Via `theme.json`, themes would register the design tokens they want for blocks:

```json
{
    "blocks": {
        "core/heading/h1": {
            "color": "var(--wp--preset--color--primary)",
            "font-size": "calc(1px * var(--wp--preset--font-size--huge))"
        },
        "core/heading/h4": {
            "color": "var(--wp--preset--color--secondary)",
            "font-size": "var(--wp--preset--font-size--normal)"
        },
        "core/paragraph": {
            "color": "var(--wp--preset--color--quaternary)",
            "font-size": "calc(1px * var(--wp--preset--font-size--normal))"
        }
    }
}
```

Note the use of CSS variables to set the values of the properties. These come from the color and font-size presets that the theme has defined in `functions.php`. They'll come from the `theme.json` at some point, but at the moment we pull them from the existing presets https://github.com/WordPress/gutenberg/pull/22076

```php
add_theme_support( 'editor-color-palette', array(
    array(
        'name'  => __( 'Primary', 'global-styles' ),
        'slug'  => 'primary',
        'color' => 'rgb(131, 12, 8)',
    ),
    array(
        'name'  => __( 'Secondary', 'global-styles' ),
        'slug'  => 'secondary',
        'color' => 'rgb(60, 12, 15)',
    ),
    array(
        'name'  => __( 'Tertiary', 'global-styles' ),
        'slug'  => 'tertiary',
        'color' => 'rgb(209, 207, 203)',
    ),
    array(
        'name'  => __( 'Quaternary', 'global-styles' ),
        'slug'  => 'quaternary',
        'color' => 'rgb(30, 30, 30)',
    ),
) );

add_theme_support( 'editor-font-sizes', array(
    array(
        'name' => __( 'Primary', 'global-styles' ),
        'size' => 16,
        'slug' => 'normal'
    ),
    array(
        'name' => __( 'Secondary', 'global-styles' ),
        'size' => 32,
        'slug' => 'big'
    ),
    array(
        'name' => __( 'Tertiary', 'global-styles' ),
        'size' => 60,
        'slug' => 'huge'
    )
) );
```

The output will be:

```css

/* These CSS Custom Properties come from the presets. */
 :root {
    --wp--preset--color--primary: rgb(131, 12, 8);
    --wp--preset--color--secondary: rgb(60, 12, 15);
    --wp--preset--color--tertiary: rgb(209, 207, 203);
    --wp--preset--color--quaternary: rgb(30, 30, 30);
    --wp--preset--font-size--normal: 16;
    --wp--preset--font-size--big: 32;
    --wp--preset--font-size--huge: 60;
    /* gradient presets from core, because the theme didn't declare any */
}

h1 {
    color: var(--wp--preset--color--primary);
    font-size: calc(1px * var(--wp--preset--font-size--huge));
}
h4 {
    color: var(--wp--preset--color--secondary);
    font-size: var(--wp--preset--font-size--normal);
}
p {
    color: var(--wp--preset--color--quaternary);
    font-size: calc(1px * var(--wp--preset--font-size--normal));
}
```

Note that, at the moment, the mapping from keys to selectors (`core` => `:root`, `core/paragraph` => `p`, etc) is hardcoded in `global-styles.php`. This needs evolving in a different PR and be taken from the block registry.
 
## How to test

1. Make sure it works:

- Enable the FSE experiment.
- Install and activate the [demo theme available here](https://github.com/nosolosw/global-styles-theme/pull/5) (or create your own taking that as an inspiration).
- Load the site in the front-end. Verify that what you see is the pattern like above.

2. Make sure it takes different values

- Modify the theme settings so the output is different. For example, in the demo theme, go to `functions.php` and modify the preset values. As an example, copy the these ones for the color palette:

```php
add_theme_support( 'editor-color-palette', array(
    array(
        'name'  => __( 'Primary', 'global-styles' ),
        'slug'  => 'primary',
        'color' => 'rgb(232, 123, 29)',
    ),
    array(
        'name'  => __( 'Secondary', 'global-styles' ),
        'slug'  => 'secondary',
        'color' => 'rgb(104, 82, 3)',
    ),
    array(
        'name'  => __( 'Tertiary', 'global-styles' ),
        'slug'  => 'tertiary',
        'color' => 'rgb(243, 236, 227)',
    ),
    array(
        'name'  => __( 'Quaternary', 'global-styles' ),
        'slug'  => 'quaternary',
        'color' => 'rgb(0, 0, 0)',
    ),
) );

```

You should see instead:

![Screenshot from 2020-05-14 11-02-08](https://user-images.githubusercontent.com/583546/81915060-61850600-95d2-11ea-912c-3dd69ee20051.png)

3) Make sure it doesn't take unsupported properties.

- Try adding an unsupported property in the `theme.json`. For example, add a text-shadow to the h1:

```
{
    "blocks": {
        "core/heading/h1": {
            "color": "var(--wp--preset--color--primary)",
            "font-size": "calc(1px * var(--wp--preset--font-size--huge))",
            "text-shadow": "2px 2px #FF0000"
        },
        "core/heading/h4": {
            "color": "var(--wp--preset--color--secondary)",
            "font-size": "var(--wp--preset--font-size--normal)"
        },
        "core/paragraph": {
            "color": "var(--wp--preset--color--quaternary)",
            "font-size": "calc(1px * var(--wp--preset--font-size--normal))"
        }
    }
}
```

- Open the dev tools of your browser of choice and verify that the `#global-styles-inline-css` embedded stylesheet doesn't contain any text-shadow property.
